### ISSUE TITLE
-W / whyrun mode doesn't work

### ISSUE THREAD

**Issue by aerickson** [boundary/bprobe_cookbook#10]
It's nice to check what a chef-client run is going to do with -W/whyrun mode... but bprobe barfs (output below). Let me know if this is totally due to a bug in apt and I'll investigate.

Thanks,
Andy

//

versions I'm running:
apt                        2.3.1
bprobe                     0.1.0

2.3.5 is out on github, but doesn't seem to have any whyrun fixes.

/////

```
 (up to date)
  * apt_repository[boundary] action add[2014-01-16T16:33:25-06:00] INFO: Processing apt_repository[boundary] action add (bprobe::dependencies line 60)

    - Would evaluate block and run any associated actions
================================================================================
Error executing action `add` on resource 'apt_repository[boundary]'
================================================================================


RuntimeError
------------
The repository file to create is nil, cannot continue.


Cookbook Trace:
---------------
/var/chef/cache/cookbooks/apt/providers/repository.rb:127:in `block in class_from_file'


Resource Declaration:
---------------------
# In /var/chef/cache/cookbooks/bprobe/recipes/dependencies.rb

 60:   apt_repository "boundary" do
 61:     uri "https://apt.boundary.com/ubuntu/"
 62:     distribution node[:lsb][:codename]
 63:     components ["universe"]
 64:     key "https://apt.boundary.com/APT-GPG-KEY-Boundary"
 65:     action :add
 66:   end
 67: end



Compiled Resource:
------------------
# Declared in /var/chef/cache/cookbooks/bprobe/recipes/dependencies.rb:60:in `from_file'

apt_repository("boundary") do
  action [:add]
  retries 0
  retry_delay 2
  cookbook_name "bprobe"
  recipe_name "dependencies" 
  uri "https://apt.boundary.com/ubuntu/"
  distribution "precise"
  components ["universe"]
  key "https://apt.boundary.com/APT-GPG-KEY-Boundary"
end



[2014-01-16T16:33:25-06:00] INFO: Running queued delayed notifications before re-raising exception
...
[2014-01-16T16:33:25-06:00] ERROR: apt_repository[boundary] (bprobe::dependencies line 60) had an error: RuntimeError: The repository file to create is nil, cannot continue.
```

stacktrace:

```
Generated at 2014-01-23 18:58:06 -0600
RuntimeError: apt_repository[boundary] (bprobe::dependencies line 60) had an error: RuntimeError: The repository file to create is nil, cannot continue.
/var/chef/cache/cookbooks/apt/providers/repository.rb:127:in `block in class_from_file'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/provider/lwrp_base.rb:138:in `instance_eval'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/provider/lwrp_base.rb:138:in `block in action'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/provider.rb:114:in `run_action'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/resource.rb:625:in `run_action'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/runner.rb:49:in `run_action'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/runner.rb:81:in `block (2 levels) in converge'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/runner.rb:81:in `each'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/runner.rb:81:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/resource_collection.rb:98:in `block in execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/resource_collection/stepable_iterator.rb:116:in `call'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/resource_collection/stepable_iterator.rb:116:in `call_iterator_block'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/resource_collection/stepable_iterator.rb:104:in `iterate'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/resource_collection.rb:96:in `execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/runner.rb:80:in `converge'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/client.rb:433:in `converge'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/client.rb:500:in `do_run'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/client.rb:199:in `block in run'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/client.rb:193:in `fork'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/client.rb:193:in `run'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/application.rb:208:in `run_chef_client'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/application/client.rb:312:in `block in run_application'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/application/client.rb:304:in `loop'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/application/client.rb:304:in `run_application'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/lib/chef/application.rb:66:in `run'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.8.2/bin/chef-client:26:in `<top (required)>'
/usr/bin/chef-client:23:in `load'
/usr/bin/chef-client:23:in `<main>'
```

**Comment by aerickson** [comment]
I think the fix is needed in the apt cookbook. It's not marking regions that modify the system with the necessary blocks  (per http://dougireton.com/blog/2013/01/07/creating-an-lwrp-part-2/).

This seems to have fixed -W with bprobe. I'll try to get this into the apt cookbook (I don't mind if you guys beat me to it though).

Andy

////

```
diff --git a/cookbooks/apt/providers/repository.rb b/cookbooks/apt/providers/repository.rb
index 0ae7bce..ce29bed 100644
--- a/cookbooks/apt/providers/repository.rb
+++ b/cookbooks/apt/providers/repository.rb
@@ -120,14 +120,17 @@ action :add do
                             new_resource.arch,
                             new_resource.deb_src)

-  file "/etc/apt/sources.list.d/#{new_resource.name}.list" do
-    owner "root"
-    group "root"
-    mode 00644
-    content repository
-    action :create
-    notifies :delete, "file[/var/lib/apt/periodic/update-success-stamp]", :immediately
-    notifies :run, "execute[apt-get update]", :immediately if new_resource.cache_rebuild
+  description = "have placed /etc/apt/sources.list.d/#{new_resource.name}.list"
+  converge_by(description) do
+    file "/etc/apt/sources.list.d/#{new_resource.name}.list" do
+      owner "root"
+      group "root"
+      mode 00644
+      content repository
+      action :create
+      notifies :delete, "file[/var/lib/apt/periodic/update-success-stamp]", :immediately
+      notifies :run, "execute[apt-get update]", :immediately if new_resource.cache_rebuild
+    end
   end
 end
```

**Comment by aerickson** [comment]
This was due to having an old version of apt loaded in by another library... floating that recipe fixed this (it was using apt 1.9.2 vs 2.3.1). You can check which version the client is using by inspecting '/var/chef/cache/cookbooks/apt/metadata.rb'.

Andy

**Comment by aerickson** [comment]
You may want to require a version past 1.9.2 up to 2.3.1/2.3.4 (not sure at what point whyrun started to work in apt basically...). I think most people expect -W/whyrun to work. If those versions were required I think in my scenario I would have gotten an error when the client attempted to reconcile the dependencies.
